### PR TITLE
Fold ifcmp with two constant integral operands

### PR DESCRIPTION
**Pull request by liqunl**
Evaluate ifcmp in walker when its operands are integral constants and
fold it. This helps reducing method size of a custom MethodHandle thunk
as code like `if (ILGenMacros.isShareableThunk())` will be folded away.

#4837 

Signed-off-by: Liqun Liu <liqunl@ca.ibm.com>

### CODE REVIEW

**Comment by liqunl** [comment]
@andrewcraik May I ask for a review?

File: runtime/compiler/ilgen/Walker.cpp [hunk 1]
```diff
@@ -3578,13 +3578,53 @@ TR_J9ByteCodeIlGenerator::genIfTwoOperand(TR::ILOpCodes nodeop)
...
 
+   TR::DataType type = first->getDataType();
+   if (branchBC > _bcIndex &&
+       first->getOpCode().isLoadConst() &&
+       second->getOpCode().isLoadConst() &&
+       type != TR::Address &&
+       type != TR::Float &&
+       type != TR::Double)
+      {
+      int64_t v1 = first->getConstValue();
```
Comment (andrewcraik) [inline runtime/compiler/ilgen/Walker.cpp]: is the widening of narrow types in this way guaranteed to always do the right thing? for equal/not equal the extension mode doesn't really matter, but for relative operations it does - do we need to respect the original type and widen 'the right way' for whatever that means?

File: runtime/compiler/ilgen/Walker.cpp [hunk 2]
```diff
@@ -3578,13 +3578,53 @@ TR_J9ByteCodeIlGenerator::genIfTwoOperand(TR::ILOpCodes nodeop)
...
+            break;
+         case TR_cmpGE:
+            branchTaken = v1 >= v2;
+            break;
+         }
+
+      if (comp()->getOption(TR_TraceILGen))
+         traceMsg(comp(), "%s\n", branchTaken ? "taking the branch" : "fall through");
+
+      return branchTaken ? branchBC : fallThruBC;
```
Comment (andrewcraik) [inline runtime/compiler/ilgen/Walker.cpp]: what about the _blocksToInline stuff below and associated gens - do we need to do that still?

Comment (liqunl) [reply runtime/compiler/ilgen/Walker.cpp]: `_blocksToInline` is used in partial inlining, we still need it. What do you mean by associated gens?

File: runtime/compiler/ilgen/Walker.cpp (hunk 1 above)
Comment (liqunl) [reply runtime/compiler/ilgen/Walker.cpp]: An integral constant, regardless of its type, is stored in `_unionBase._constValue` which is a `int64_t`.  Any conversion has been done when storing the constant to `_unionBase._constValue`. But there will be problem in unsigned comparison that I'll fix.

File: runtime/compiler/ilgen/Walker.cpp (hunk 2 above)
Comment (andrewcraik) [reply runtime/compiler/ilgen/Walker.cpp]: You are returning earlier than the follow-on code - if partial inlining is on or some other situation do we need to run the code below this point - it seems surprising that it is an unconditional return early.

Comment (liqunl) [reply runtime/compiler/ilgen/Walker.cpp]: The partial inlining code seems complicated to deal with, I don't know how it supposed to work and I have never seen partial inlining kick in. I added a condition to only fold the if if we're not in partial inlining.

Comment (andrewcraik) [reply runtime/compiler/ilgen/Walker.cpp]: where is this check?

Comment (liqunl) [reply runtime/compiler/ilgen/Walker.cpp]: It is at line 3589.

Comment (andrewcraik) [reply runtime/compiler/ilgen/Walker.cpp]: Wouldn't it be better to move that check down near the return and log a failure message about not folding due to partial inlining so it is easier to find this later if we need to fix it?

Comment (liqunl) [reply runtime/compiler/ilgen/Walker.cpp]: Moved the check down

**Comment by andrewcraik** [comment]
Jenkins test sanity xlinux,win,plinux jdk8,jdk11

**Pull request merged**